leader-election/src/main/java/com/iluwatar/leaderelection/AbstractInstance.java

```diff
@@ -0,0 +1,153 @@
+/*
+ * The MIT License
+ * Copyright © 2014-2019 Ilkka Seppälä
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+
+package com.iluwatar.leaderelection;
+
+import java.util.Queue;
+import java.util.concurrent.ConcurrentLinkedQueue;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Abstract class of all the instance implementation classes.
+ */
+public abstract class AbstractInstance implements Instance, Runnable {
+
+  private static final Logger LOGGER = LoggerFactory.getLogger(AbstractInstance.class);
+
+  protected static final int HEARTBEAT_INTERVAL = 5000;
+
+  protected MessageManager messageManager;
+  protected Queue<Message> messageQueue;
+  protected final int localId;
+  protected int leaderId;
+  protected boolean alive;
+
+  /**
+   * Constructor of BullyInstance.
+   */
+  public AbstractInstance(MessageManager messageManager, int localId, int leaderId) {
+    this.messageManager = messageManager;
+    this.messageQueue = new ConcurrentLinkedQueue<>();
+    this.localId = localId;
+    this.leaderId = leaderId;
+    this.alive = true;
+  }
+
+  /**
+   * The instance will execute the message in its message queue periodically once it is alive.
+   */
+  @Override
+  @SuppressWarnings("squid:S2189")
+  public void run() {
+    while (true) {
+      if (!this.messageQueue.isEmpty()) {
+        this.processMessage(this.messageQueue.remove());
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.



**Problem**  
 You are using a `ConcurrentLinkedQueue`, but your usage of `isEmpty()` and `remove()` may not be thread-safe at lines: **64 and 65**. If the `ConcurrentLinkedQueue` contains only **one element** and two threads perform this **same check at the same time** , the second thread&#39;s call to `remove()` will throw `NoSuchElementException`.

**Fix**  
 Consider using `poll()` (does not throw an exception, but it may return `null`, which you can check for) or handling the exception.